Post Comments block: hide the "Comments Closed" message

jffng opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
Solves https://github.com/WordPress/gutenberg/issues/35732. 

This PR changes the Post Comments rendering to hide the "Comments are closed" message when the following conditions are met:

- Comments are closed
- There are no comments to display
  
## How has this been tested?

1. Check out this PR: https://github.com/WordPress/twentytwentytwo/pull/127
2. Ensure "Allow comments" is disabled on the post, and no comments exist
3. Verify that no post comments markup is rendered
4. Go to a post with comments and verify they still appear

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
New feature (non-breaking change which adds functionality)
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

jffng committed:
Add hideCommentsClosed attribute.

jffng committed:
Apply css class instead to hide "Comments closed" message

jffng committed:
Remove default value.

jffng committed:
Make php linter happy.

scruffian committed:
Check that the attribute is not empty before we check its value

scruffian commented:
I was seeing a notice when the attribute wasn't set. I added a commit to fix that.

SantosGuillamot commented:
I've been taking a look and it seems to be working fine! I think I am not the best person to review the code, but I checked how it works in the editor 🙂 .

> Note that no UI is added in this PR, since I am not sure this approach aligns with the work being doing on the Comments Template block #35000.

Regarding this, there is still not progress on the Comments Template block, but this is for sure something to consider. I will post it there so it isn't missed.

aristath commented on packages/block-library/src/post-comments/index.php:
```diff
@@ -38,6 +38,9 @@ function render_block_core_post_comments( $attributes, $content, $block ) {
 	}
+	if ( ! comments_open( $post_id ) && ! empty( $attributes['hideCommentsClosed'] ) && $attributes['hideCommentsClosed'] ) {
+		$classes .= 'hide-comments-closed';
+	}
```
Would it not be more efficient to early return instead of adding an extra class? 🤔 
This way if the post doesn't have comments and we want to hide comments-closed, no HTML will be rendered (as opposed to printing it and then hide it via CSS)

```suggestion
	if ( ! comments_open( $post_id ) && ! empty( $attributes['hideCommentsClosed'] ) && $attributes['hideCommentsClosed'] ) {
		$args = array(
			'post_id' => $post_id,
			'count'   => true,
		);
		if ( ! get_comments( $args ) ) {
			return;
		}
	}
```

jffng replied:
I considered an early return, but decided against it because the message would still show if there are comments. 

If we're okay only hiding the message if there are no comments, then I'd much prefer this approach!

kjellr replied:
I think that’s fine. The issue is just that the message shows when there are no comments _and_ comments are closed.

kjellr replied:
But actually… should that just be the default behavior of the block then? Rather than something that’s opted into?

- If comments are closed, show the message. 
- If comments are closed *and* there are no comments, don’t show the message.


I can’t think a reason why you would want this message showing up under that second scenario. 🤔

jffng committed:
Update packages/block-library/src/post-comments/index.php

Co-authored-by: Ari Stathopoulos <aristath@gmail.com>

jffng committed:
Remove unneeded CSS.

jffng commented:
Updated the PR based on the suggestion and logic described [here](https://github.com/WordPress/gutenberg/pull/35743#discussion_r733513542), so no CSS is needed. I also updated the original PR description. It should be ready for another review.

jffng commented:
Another thought — what if we got rid of the attribute altogether, and just returned early if comments are closed and there are no comments.

kjellr commented:
> Another thought — what if we got rid of the attribute altogether, and just returned early if comments are closed and there are no comments.

Yeah, this seems like the right move to me.

jffng committed:
Remove attribute.

jffng committed:
Fix linting errors.

aristath reviewed (approved):
Thank you for the changes!
LGTM 👍

kjellr reviewed (approved):
This works well in my testing too. Should be good to merge once tests pass. 👍
